File: merge_sort/merge_sort.py
def merge(array, left, right):
    """
    Perform Merge Operation between arrays.
    Time Complexity: Theta(nLogn)
    Auxiliary Space: O(n)
    :param array: Iterable of elements
    :param left: left limit for merge sort
    :param right: right limit for merge sort
    :return: no returns, merges arrays.
    """
    mid = (left + right) // 2
    l = array[left:mid + 1]
    r = array[mid + 1:right + 1]
    k = left
    while l and r:
        if l[0] < r[0]:
            array[k] = l.pop(0)
        else:
            array[k] = r.pop(0)
        k += 1
    while l:
        array[k] = l.pop(0)
        k += 1
    while r:
        array[k] = r.pop(0)
        k += 1


def merge_sort(array, left, right):
    """
    Perform sort using merge function.
    Time Complexity : O(nlog(n)).
    Space Complexity : O(n).
    :param array: Iterable of elements
    :param left: left limit for merge sort
    :param right: right limit for merge sort
    :return: no returns, sorts array
    """
    if left < right:
        mid = (left + right) // 2
        merge_sort(array, left, mid)
        merge_sort(array, mid + 1, right)
        merge(array, left, right)


def main():
    a = [15, 19, 18, 26, 456, 87, 45, -1, 558897984]
    merge_sort(a, 0, len(a) - 1)
    print(a)


if __name__ == '__main__':
    main()

File: merge_sort/merge_sort.c
/*
 *  Following Algorithm sorts the given array in worst case time complexity O(n*log(n))
 *  and space complexity O(n), n being size of array.
 */

#include <stdio.h>

void merge(int *arr, int left, int mid, int right) {
	int i, j, k;
	int n_left = mid - left + 1;
	int n_right = right - mid;
	int left_arr[n_left];
	int right_arr[n_right];
	for (i = 0; i < n_left; i++)
		left_arr[i] = arr[left + i];
	for (j = 0; j < n_right; j++)
		right_arr[j] = arr[mid + 1 + j];
	i = 0;
	j = 0;
	k = left;
	while (i < n_left && j < n_right) {
		if (left_arr[i] <= right_arr[j]) {
			arr[k] = left_arr[i];
			i++;
		} else {
			arr[k] = right_arr[j];
			j++;
		}
		k++;
	}
	while (i < n_left) {
		arr[k] = left_arr[i];
		i++;
		k++;
	}
	while (j < n_right) {
		arr[k] = right_arr[j];
		j++;
		k++;
	}
}

void merge_sort(int *arr, int left, int right) {
	if (left < right) {
		int mid = (left + right) / 2;
		merge_sort(arr, left, mid);
		merge_sort(arr, mid + 1, right);
		merge(arr, left, mid, right);
	}
}

int main() {
	int i;
	int arr[] = {7, 10, 14, 1, 2, 6, 17, 0};
	int size = sizeof(arr) / sizeof(arr[0]);
	merge_sort(arr, 0, size - 1);
	for (i = 0; i < size; i++) {
		printf("%d ", arr[i]);
	}
	return 0;
}
